dali/imgcodec/parsers/jpeg.cc

```diff
@@ -13,18 +13,65 @@
 // limitations under the License.
 
 #include "dali/imgcodec/parsers/jpeg.h"
+#include "dali/core/byte_io.h"
 
 namespace dali {
 namespace imgcodec {
 
+using jpeg_marker_t = std::array<uint8_t, 2>;
+
+constexpr jpeg_marker_t sos_marker = {0xff, 0xda};
+constexpr jpeg_marker_t soi_marker = {0xff, 0xd8};
+constexpr jpeg_marker_t eoi_marker = {0xff, 0xd9};
+
+bool isValidMarker(const jpeg_marker_t &marker) {
+  return marker[0] == 0xff && marker[1] != 0x00;
+}
+
+bool isSofMarker(const jpeg_marker_t &marker) {
+  if (!isValidMarker(marker) || marker[1] < 0xc0 || marker[1] > 0xcf) return false;
+  return (marker[1] != 0xc4 && marker[1] != 0xc8 && marker[1] != 0xcc);
+}
+
 ImageInfo JpegParser::GetInfo(ImageSource *encoded) const {
-  ImageInfo info;
-  // info.shape = {h, w, c};   // TODO(janton)
+  ImageInfo info{};
+  auto stream = encoded->Open();
+
+  jpeg_marker_t first_marker = stream->ReadOne<jpeg_marker_t>();
+  DALI_ENFORCE(first_marker == soi_marker);
+
+  bool read_shape = false;
+  while (!read_shape) {
+    auto marker = stream->ReadOne<jpeg_marker_t>();
+    if (marker[1] == 0xff) {
+      stream->SeekRead(-1, SEEK_CUR);
+      continue;
+    }
+
+    DALI_ENFORCE(isValidMarker(marker));
+    if (marker == sos_marker) break;
```

Comment:
(Just a suggestion) How about something like this:
```
  jpeg_marker_t marker;
  marker[0] = stream->ReadOne<uint8_t>();
  DALI_ENFORCE(marker[0] == 0xFF, make_string("A valid marker must start with 0xFF, got: ", marker[0]));
  // https://www.w3.org/Graphics/JPEG/itu-t81.pdf section B.1.1.2 Markers
  // Any marker may optionally be preceded by any number of fill bytes, 
  // which are bytes assigned code X’FF’ */
  do {
    marker[1] = stream->ReadOne<uint8_t>();
  } while (marker[1] == 0xFF);
  DALI_ENFORCE(isValidMarker(marker), make_string("Invalid marker: ", marker));
  if (marker == sos_marker) break;
```

Reply:
Good idea, that's much cleaner. I've applied most of this in https://github.com/NVIDIA/DALI/pull/4050/commits/539344ff02f9df8ca056a40b66f65f712e8c7b9b. 

The only problem I see is that `make_string` won't pretty-print `uint8` as hex, but will just interpret the number as ASCII code, resulting in some trash being printed. As far as I understand the easiest to nicely print it as hex would be with some `ostringstream` and I doubt if it's worth it.

Reply:
Probably not worth the effort